### ISSUE TITLE
Define and implement a REST interface for the Credentials API

### ISSUE THREAD

**Issue by sysexcontrol** [eclipse/hono#350]
The credentials API currently consists of an AMQP interface that is split into mandatory (GET operation) and optional parts (ADD,UPDATE,REMOVE operations).

The mandatory GET operation is already used by the MQTT adapter for allowing a setup where device need to authenticate themselves by username/password.

However, the administration is currently purely done by editing the credentials file and restarting the device registry component.

To allow the administration of the credentials we thus need a REST interface for all 4 operations
ADD, GET, UPDATE, REMOVE - like we have for the Device Registration API already.

Remark: a first version of the ADD operation is already provided in the code base.

**Comment by ppatierno** [comment]
Initial add operation provided by 62922e957586475ac161a4df826ce359e18a67d7

**Comment by sysexcontrol** [comment]
@ppatierno @sophokles73 : 
The currently reached state of the implementation is available in master branch and consists of:

- POST plus full tests, DELETE plus full tests, GET plus almost full tests (one TODO).
- PUT is missing so far.
- Full documentation of these requests at the same place as the registration REST API was documented (under the component documenation of device registry).

I labeled the documentation as DRAFT so far.

**There is one point that I like to discuss regarding the RESTful-ness of the API:**
Look at the URIs that now are available:

`POST /credentials/tenantId with payload`
`DELETE /credentials/tenantId/deviceId[/type/][authId]`
`GET /credentials/tenantId/authId[/type]`

The GET URI is just like a protocol adapter would need it for authentication (as already implemented for the AMQP version of the credentials API).

But for RESTful handling it might be better like

`GET /credentials/tenantId/deviceId/[type]/[authId]` to stay with the same resources as DELETE (and PUT when avail).

So basically I have two questions in mind:

1. Should we use the GET like `"/credentials/tenantId/deviceId[/type][/authId]"` and skip the other (now implemented) variant? Meaning that protocol adapters stick to the AMQP version of GET (which would be the only operation then)?
2. Or should we have the GET in two variants: one for the protocol adapters (under a new endpoint like `"/authentication/tenantId/authId/type"` then), and one for the RESTful handling of credentials (under the endpoint `"/credentials/tenantId/deviceId[/type][/authId]"`)?

My preference would be 2.

WDYT?

**Comment by sophokles73** [comment]
GET should use the URI returned in the *Location* header when adding credentials.
The following should be possible (using auth-id `sensor1`):
1. POST /credentials/tenant
    *hashed password credentials payload*
     
    201 Created
    Location /credentials/tenant/sensor1/hashed-password
1. GET  /credentials/tenant/sensor1/hashed-password
    
    200 Ok
    *payload*
1. DELETE /credentials/tenant/sensor1/hashed-password
    
    204 No Content

In addition, we can have a resource `DELETE /credentials/tenant` which accepts a request payload containing the query params defined by the Credentials API to remove multiple credentials at once.

**Comment by sophokles73** [comment]
I have also added a resource for updating existing credentials:
PUT /credentials/${tenant}/${auth-id}/${type}

**Comment by sophokles73** [comment]
fixed in 1fbb7696864875cbf15662e9d79501d764c52bfb
